_posts/2017-12-31-3-stages-api-testing.md

```diff
@@ -0,0 +1,427 @@
+
+# Stage 1: Unit testing the communication layer (front and back-end)
+
+<p>
+    <img class="image fit" style="max-width:575px" alt="decoupled" src="/img/3-stages-api-testing/unit.png" />
+</p>
+
+The first stage is unit testing each component. This step is almost always part of the component's build. Let's take a quick look at each component.
+
+### Unit testing the communication part of the UI
+
+<p>
+    <img class="image fit" style="max-width:325px" alt="Easy client server setup" src="/img/3-stages-api-testing/unit-client.png" />
+</p>
+
+For unit testing our front-end Angular 5 application, we are going to use the `import { MockBackend } from '@angular/http/testing';` from Angular itself to mock our back-end.
+
+```
+...
+beforeEach(async(() => {
+    TestBed.configureTestingModule({
+        ...
+      providers: [
+        CCService, // our service that is handling the communication
+        {provide: XHRBackend, useClass: MockBackend} // our mock
+      ]
+    })
+    .compileComponents();
+  }));
+...
+```
+
+```
+describe('Should query ccs with an observable', () =>
+    it('Should return data', inject([CCService, XHRBackend], (ccService, mockBackend) => {
+      mockBackend.connections.subscribe((connection) => {
+        const ccs: Array<CC> = [
+          {
+              ...
+          },
+          {
+              ...
+          }
+        ];
+        connection.mockRespond(new Response(new ResponseOptions({
+          body: JSON.stringify(ccs)
+        })));
+      });
+      ccService.getStats('tim').subscribe((ccs: Array<CC>) => {
+        expect(ccs.length).toBe(2);
+        expect(ccs[0].clEmail).toBe('tim.vierbergen@ordina.be');
+       });
+      })
+    )
+  );
+```
+
+The Angular framework handles the communication part.
+We want to mock the `$http` call and see how we are handling the result.
+The `mockBackend` is working inbetween our own code and the provided `$http` part.
+Every call is triggering the `mockBackend.connections`, so the subscribers are triggered.
+We are providing our own data and returning it as the response of the `$http` call.
+
+### Unit testing the communication layer of the service
+
+<p>
+    <img class="image fit" style="max-width:325px" alt="Easy client server setup" src="/img/3-stages-api-testing/unit-service.png" />
+</p>
+
+For unit testing our back-end, we are using Jest.
+<a href="https://facebook.github.io/jest/" target="_blank">Jest</a> is a testing framework by Facebook.
+If you are interested in Jest, make sure to watch this blog, as a post around this topic is in the making.
+Just as with our front-end, we trust the framework to correctly handle the communication itself.
+
+Our unit tests will start at the controller level of our API.
+
+```
+describe('searchCCs', () => {
+    it('should return a filtered array of competence centers', async () => {
+      await ccsController.searchCCs(mockResponse, 'tim');
```

Comment:
The async searchCCs is not returning anything. You could argue this is bad practice, but it's implemented like this at the moment. Therefore the test itself is using the Mockresponse (that is getting prepared to send back) itself.
```
async searchCCs(@Response() res, @Query('needle') searchString) {
  let filtered = this._ccsService.searchCCs(searchString);
  res.status(HttpStatus.OK).json(filtered);
}
```
When doing the `let response = await ccsController.searchCCs(mockresponse, 'time);` response === mockresponse because that is the actual object we are going to send back.

And yes, we are just waiting for the search to be finished so (mock)response.data is filled with the data. `this._ccsService.searchCCs` can, for example, call databases and calculate data...